### ISSUE TITLE
`test_unknown_key` may crash ssh-agent

### ISSUE THREAD

**Issue by vincentbernat** [ronf/asyncssh#114]
With OpenSSH 7.6p1, `test_unknown_key()` will crash the SSH agent.

```
test_unknown_key (tests.test_agent._TestAPI)
Test failure when signing with an unknown key ... debug2: fd 4 setting O_NONBLOCK
debug1: process_message: socket 1 (fd=4) type 19
debug1: process_message: socket 1 (fd=4) type 13
process_sign_request2: buffer error: invalid format
debug1: cleanup_socket: cleanup
[1]  + 26427 exit 255   ssh-agent -a /tmp/agent -d
```

The code in `ssh-agent` leading to the crash (calling `fatal()`) is not recent, so I don't know why this happens only now and why this worked in the past. Maybe, the test could be corrected by providing a valid key.

**Comment by ronf** [comment]
Thanks very much for the additional detail to help track this down! It's odd, though - a buffer error shouldn't occur with an invalid key. It should only occur if the sign request is malformed (missing one of its arguments), or if an error of some kind occurs when constructing the buffer containing the response.

Can you confirm which "buffer error" line you are hitting here, and what the e->request looks like before it attempts to parse it? I'll also try to reproduce it here now I have more details.

The goal of the test is to trigger an error response from the agent, so providing an invalid key is really the point. I could provide a key which is valid but not present in the agent to trigger a different failure path there, but neither one of those paths should lead to a buffer error. The one I'm using now should trigger:

                error("%s: cannot parse key blob: %s", __func__, ssh_err(r));

and a key which is valid but not loaded into the agent should trigger:

                verbose("%s: %s key not found", __func__, sshkey_type(key));

Neither of these is a fatal().

Also, I just noticed that I seem to be unintentionally repeating this unknown key test down in test_errors(). Does that ever lead to the same crash? There has to be something else going on here which is causing it to generate an invalid request message.

**Comment by vincentbernat** [comment]
I have a hard time to use ssh-agent under gdb. After hitting any breakpoint, I get this error:

```
Cannot remove breakpoints because program is no longer writable.
Further execution is probably impossible.

Breakpoint 1, process_message (socknum=1) at ../../ssh-agent.c:709
709     ../../ssh-agent.c: No such file or directory.
```

So, I am not able to get a backtrace. Which version of ssh-agent are you using on your side?

In `test_errors()`, the test doesn't crash, but it's using a mock agent, isn't it?

**Comment by ronf** [comment]
Unfortunately, line 709 in ssh_agent.c is the call to process_sign_request2(). To figure out which fatal() it was, we'd need one stack frame down from there inside process_sign_request2().

Regarding test_errors(), you're right about the mock agent - I actually missed that when I was looking through the code earlier. So, that's not relevant here.

If I can reproduce this, I think the easiest thing will probably be to actually insert some additional debug messages in the SSH agent itself, but it might also be possible to do this with a debugger if you download the source code and make sure gdb can find it. To print some of the variables, though, you might need to recompile the code with debugging enabled. I doubt the Debian packages would be compiled with debugging.

Unfortunately, I won't get a chance to dig into this today, but I'll try and take a closer look at this during the week.

**Comment by vincentbernat** [comment]
I have debug symbols (package openssh-client-dbgsym) and I am able to set the breakpoints to `fatal()` or `process_sign_request2()` but once the breakpoint is hit, I get the error above and I cannot step or examine memory. Backtrace is not available either. First time I get that and googling doesn't help.

**Comment by ronf** [comment]
I took a closer look at this tonight on MacOS, and was able to reproduce this with OpenSSH 7.6p1 in a simple test. Calling agent.sign(b'xxx', b'test') generated the following output in 7.6p1:

```
debug2: fd 3 setting O_NONBLOCK
debug3: fd 4 is O_NONBLOCK
debug1: process_message: socket 1 (fd=4) type 13
process_sign_request2: buffer error: invalid format
debug1: cleanup_socket: cleanup
Process 92688 exited with status = 255 (0x000000ff) 
```

However, running the exact same test with 7.5p1 produced:

```
debug2: fd 3 setting O_NONBLOCK
debug3: fd 4 is O_NONBLOCK
debug1: type 13
process_sign_request2: cannot parse key blob: invalid format
debug1: XXX shrink: 3 < 4
```

Note that the error here is "cannot parse key blob", which is an error() call, not a fatal(). That's what I would expect to get here when I pass in a bad key blob, and that properly returns a failure and leaves the agent running.

So, there's definitely a regression of some sort that is specific to 7.6p1. Looking more closely at the code, I think I see the problem. In 7.5p1, the code looks like:

```c
        if ((r = sshbuf_get_string(e->request, &blob, &blen)) != 0 ||
            (r = sshbuf_get_string(e->request, &data, &dlen)) != 0 ||
            (r = sshbuf_get_u32(e->request, &flags)) != 0)
                fatal("%s: buffer error: %s", __func__, ssh_err(r));
        if (flags & SSH_AGENT_OLD_SIGNATURE)
                compat = SSH_BUG_SIGBLOB;
        if ((r = sshkey_from_blob(blob, blen, &key)) != 0) {
                error("%s: cannot parse key blob: %s", __func__, ssh_err(r));
                goto send;
        }
```

However, in 7.6p1, this changed to:

```c
        if ((r = sshkey_froms(e->request, &key)) != 0 ||
            (r = sshbuf_get_string_direct(e->request, &data, &dlen)) != 0 ||
            (r = sshbuf_get_u32(e->request, &flags)) != 0)
                fatal("%s: buffer error: %s", __func__, ssh_err(r));
```

Switching from sshbuf_get_string() with a separate call to sshkey_from_blob() to a call to sshkey_froms() is causing any invalid key blob to be treated as a fatal parsing error of the input message that causes the agent to exit, instead of being handled as a non-fatal error in parsing the key which returns an error.

I can probably work around this for now by passing in a key blob which is valid but not currently loaded. In both versions, there's a later call to:

```c
        if ((id = lookup_identity(key, 2)) == NULL) {
                verbose("%s: %s key not found", __func__, sshkey_type(key));
                goto send;
        }
```

This should result in a failure (though for a different reason than before) and avoid the race we're seeing with "No such process" in the AsyncSSH unit tests when cleaning up. OpenSSH exiting out of the agent just because it can't parse a key seems like a bad idea, though, so I'll probably file a bug in OpenSSH about this as well. I think the old code was more robust.

**Comment by ronf** [comment]
Here's a potential workaround for this issue:

```patch
diff --git a/tests/test_agent.py b/tests/test_agent.py
index 966d839..b75fda2 100644
--- a/tests/test_agent.py
+++ b/tests/test_agent.py
@@ -348,8 +348,10 @@ class _TestAPI(AsyncTestCase):
     def test_unknown_key(self, agent):
         """Test failure when signing with an unknown key"""
 
+        key = asyncssh.generate_private_key('ssh-rsa')
+
         with self.assertRaises(ValueError):
-            yield from agent.sign(b'xxx', b'test')
+            yield from agent.sign(key.get_ssh_public_key(), b'test')
 
     @agent_test
     def test_double_close(self, agent):
```

It's getting late, so I won't be checking this in tonight, but if all goes well I should be able to check this in to the "develop" branch tomorrow. I want to see if I can trigger the error without the fix more reliably, perhaps by putting in a sleep before it tries to clean up the ssh-agent process to give it time to exit. While I've confirmed in a manual test that the agent now reports "RSA key not found" and doesn't exit with this change, seeing the "No such process" error without the fix will help to close the loop here.

**Comment by ronf** [comment]
This change is now checked into the "develop" branch.

I also confirmed that without this change I could reliably reproduce this issue by adding a short sleep before the os.kill() call in the unit test teardown. It was sporadic because the kill sometimes ran before the SSH agent had finished exiting on its own.

I also reported the issue on the OpenSSH developer mailing list, and this will be fixed in an upcoming OpenSSH release, now reporting an error for malformed arguments in a sign request, rather than exiting with a fatal error in those cases.
